apisix/plugins/api-breaker.lua

```diff
@@ -0,0 +1,214 @@
+                }
+            }
+        }
+    },
+    required = {"unhealthy_response_code", "unhealthy", "healthy"},
+}
+
+
+local function is_unhealthy(unhealthy_status, upstream_status)
+    for _, unhealthy in ipairs(unhealthy_status) do
+        if unhealthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function is_healthy(healthy_status, upstream_status)
+    for _, healthy in ipairs(healthy_status) do
+        if healthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function healthy_cache_key(ctx)
+    return "healthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_cache_key(ctx)
+    return "unhealthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_lastime_cache_key(ctx)
+    return "unhealthy-lastime" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local _M = {
+    version = 0.1,
+    name = plugin_name,
+    priority = 1005,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.access(conf, ctx)
+    local unhealthy_val, err = shared_buffer:get(unhealthy_cache_key(ctx))
+    if err then
+        core.log.error("failed to get unhealthy_cache_key in ngx.shared:", err)
+    end
+
+    local unhealthy_lastime, err = shared_buffer:get(unhealthy_lastime_cache_key(ctx))
+    if err then
+        core.log.error("failed to get unhealthy_lastime_cache_key in ngx.shared: ", err)
+    end
+
+    if unhealthy_val and unhealthy_lastime then
+        local ride = math.ceil(unhealthy_val / conf.unhealthy.failures)
+        if ride < 1 then
+            ride = 1
+        end
+
+        -- The maximum intercept request is 5 minutes(DEFAULT_EXPTIME),
+        -- and then the upstream service will be retry.
+        if unhealthy_lastime + 2^ride >= ngx.time() then
+            return conf.unhealthy_response_code
+        end
+    end
+end
+
+
+function _M.header_filter(conf, ctx)
+    local unhealthy_status = conf.unhealthy.http_statuses
+    local healthy_status = conf.healthy.http_statuses
+
+    local unhealthy_key = unhealthy_cache_key(ctx)
+    local healthy_key = healthy_cache_key(ctx)
+
+    local upstream_status = core.response.get_upstream_status(ctx)
+
+    if is_unhealthy(unhealthy_status, upstream_status) then
+        local newval, err = shared_buffer:incr(unhealthy_key, 1, 0, DEFAULT_EXPTIME)
+        if err then
+            core.log.error("failed to incr unhealthy_key in ngx.shared: ", err)
+        end
+        shared_buffer:expire(unhealthy_key, DEFAULT_EXPTIME)
+        shared_buffer:delete(healthy_key) -- del healthy numeration
+
+        if 0 == newval % conf.unhealthy.failures then
+            shared_buffer:set(unhealthy_lastime_cache_key(ctx), ngx.time(), DEFAULT_EXPTIME)
```

Comment:
do we have any test case to cover this logic?

Reply:
Already have, tested the state change.